### ISSUE TITLE
Dataset of varying sizes

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#155]
To make the best model that generalizes to creting level segments of any size, we need a slightly more sophisticated dataset. Once you have closed issue #96 (which you should be able to do soon), work on this.

I believe that the diffusion model in our code should generalize to arbitrary sizes because it is convolutional. Therefore, I think it can train on a dataset that consists of mixed sizes. However, it is still important that all samples in a given batch be the same size, and because we have three down/up layers that differ by multiples of 2, it is also important that we can divide sizes by 2 at least 3 times. Therefore, I would like a dataset like the following:

Sliding window of all 16 by 16 scenes.
Sliding window of all 32 by 16 scenes.
Sliding window of all 64 by 16 scenes.
Sliding window of all 128 by 16 scenes.

In other words, take data sets created with widths for each power of 2 from 16 and up, and combine them into one.

Some special code will need to be added to make sure that the dataset/dataloader recognizes these differences, and maintains samples of the same sizes in the same buckets, so that the model only ever sees samples of a consistent size in any batch. Once you have the dataset itself and the code for bucketing the data, you can train a model (but you should also post the steps for preparing the data and training it here so that I can repeat the process on my faster home machine).

By the way, a lot of the ideas for this came from this Claude thread which might be useful to reference:
https://claude.ai/share/86206051-513f-444e-a1c2-7a83e41b3de1
There are some other ideas in here that might be worth pursuing later, specifically related to the positional encodings, but let's try training on data without that extra complication first.

**Comment by fwilson12** [comment]
After implementing the code in the most recent commit, I began training an unconditional model on a fully granular dataset consisting of levels of width 16, 32, 64, and 128. So far it's on epoch 15 and hasn't crashed, will update the thread with results as the training continues

**Comment by schrum2** [comment]
Make it so that the samples that are periodically generated during training explicitly represent one sample from each bucket (each different size) whenever multiple sizes are being used (but work as normal otherwise). With this change in place, try training another unconditional model, and once that is done, train a text conditional model.

**Comment by fwilson12** [comment]
Stopped training after 440 epochs, and the results are very promising.

Not only was the model able to produce good-looking levels of the same sizes it was trained on(16, 32, 64, 128); it was also able to generalize to scene sizes it wasn't trained on, as seen below.

## **16x16**


<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/be48de38-d77f-4ede-9697-f26634ea88ac" />




<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/3ab6caa5-5a70-45de-b983-0a12220f13be" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/1492c192-5f80-4495-b94c-ec5dbe1d777a" />



## **32x16**



<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/6c5cd249-8f47-4e5e-9045-c9901fe85d06" />

<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/d6bc51cb-29e4-44e7-8b6e-28203374b5c0" />

<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/e66f66cc-d8b9-450f-b2cf-f11ae84ca2b8" />

## **64x16**

<img width="1024" height="256" alt="Image" src="https://github.com/user-attachments/assets/093404d2-9ca3-4140-8128-9433bbd88360" />


<img width="1024" height="256" alt="Image" src="https://github.com/user-attachments/assets/0ae3b1d3-3831-4e62-a00c-e00aeb8e3853" />

<img width="1024" height="256" alt="Image" src="https://github.com/user-attachments/assets/53d78aeb-36fa-4726-8bbb-944aa7fe15e1" />


## **128x16**


<img width="2048" height="256" alt="Image" src="https://github.com/user-attachments/assets/96c6d46c-c198-4a9d-a1bc-97f481b7945e" />

<img width="2048" height="256" alt="Image" src="https://github.com/user-attachments/assets/cb80b08e-00fa-4acb-bdc8-d97064be86e1" />

<img width="2048" height="256" alt="Image" src="https://github.com/user-attachments/assets/83ec2312-e130-4b6c-9a01-1d03f4794ab4" />




## **24x16 *(not trained on)* **

<img width="384" height="256" alt="Image" src="https://github.com/user-attachments/assets/1b3565d8-f35d-4f05-a0e8-ecc42d1b2766" />

<img width="384" height="256" alt="Image" src="https://github.com/user-attachments/assets/3b68c153-1b38-4544-a8a4-d2f14b2ffc65" />

<img width="384" height="256" alt="Image" src="https://github.com/user-attachments/assets/964c3eac-8097-4eeb-ab87-12d7c49aa598" />


## **80x16 *(not trained on)* **





<img width="1280" height="256" alt="Image" src="https://github.com/user-attachments/assets/9705e927-e01e-4c85-975c-78fec3f76011" />

<img width="1280" height="256" alt="Image" src="https://github.com/user-attachments/assets/b0b84caf-097d-45a4-9fc2-65bfc61d291b" />

<img width="1280" height="256" alt="Image" src="https://github.com/user-attachments/assets/70226ed3-be6a-4a20-9ba9-cf012d8890c1" />

**Comment by fwilson12** [comment]
Next step is to train a conditional model on the same dataset, will start that now

**Comment by schrum2** [comment]
We just discussed this, but I wanted to make a note that we will train the new unconditional model for fewer epochs (we'll try 300 first). In general, since we are training on more data, it makes sense that we could train for fewer epochs, especially since the smaller samples are all slices of data that are present in the larger samples. Overfitting seems more likely, so training for less is a good idea, and has the benefit of saving us some time.

Note that 300 epochs might still be too long. We'll tune this a bit later.

**Comment by schrum2** [comment]
Add support for bucketing levels in terms of width _and_ height.

**Comment by schrum2** [comment]
Training over night with a conditional model on my home machine allowed me to reach 100 epochs, but I decided to terminate the training, since it seems to have reached a reasonable end point. The model is on Google Drive, and I am testing it more now more thoroughly, but here are the results from training:

Samples from epoch 100:

<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/c61ea085-6cd9-4ebc-96d2-84af93896c57" />
<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/5a298cb8-389a-48cd-809f-69f0ac0d391a" />
<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/3fc0a424-0ca1-49fe-9d67-98663d6dd5bb" />
<img width="512" height="256" alt="Image" src="https://github.com/user-attachments/assets/a77efda3-ed28-4334-8080-fd8f2c8bb2ae" />

I guess this was still using the setting to only generate things at the same sample size, rather than the one that generates samples at multiple sizes.

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/d42dd1d2-a18f-42aa-a41f-3009a5d5ae76" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/107b0bff-9ec4-4cfd-b552-e2887ef95d68" />

This takes advantage of a recent feature added by Alaa that also plots according to the number of steps (number of presented training batches) and not just the number of epochs. This shows that the number of steps is already way more than what we experience with the smaller training set despite the small number of epochs, which is why I terminated training.

I'm currently running evaluate_caption_adherence_multi.bat on the model and will post some results from that later, but overall this seems to work, and I guess I'll close the issue after posting those additional results.

**Comment by schrum2** [comment]
evaluate_caption_adherence_multi.bat generated some samples but also crashed when trying to process samples from Mar1and2_16-32-64-128_RandomTest-regular.json because no such file exists, but we should have a way of testing these models with both random captions and random sizes, which is related to issue #160.

Similarly, the model evaluated caption adherence on the dataset Mar1and2_16-32-64-128, but it only generated them at size 16 by 16. The samples look good, but out automated results processing should consider generating levels at different sizes now. I'll make another issue for this, but I'm going to close this one.
